plugin/health_check.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package plugin
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/mattermost/mattermost-server/mlog"
+)
+
+const (
+	HEALTH_CHECK_INTERVAL         = 30 // seconds. How often the health check should run
+	HEALTH_CHECK_DISABLE_DURATION = 60 // minutes. How long we wait for num fails to incur before disabling the plugin
+	HEALTH_CHECK_PING_FAIL_LIMIT  = 3  // How many times we call RPC ping in a row before it is considered a failure
+	HEALTH_CHECK_RESTART_LIMIT    = 3  // How many times we restart a plugin before we disable it
+)
+
+type PluginHealthCheckJob struct {
+	cancel    chan struct{}
+	cancelled chan struct{}
+	env       *Environment
+}
+
+type PluginHealthStatus struct {
+	Crashed        bool
+	failTimeStamps []time.Time
+	lastError      error
+}
+
+// InitPluginHealthCheckJob starts a new job if one is not running and is set to enabled, or kills an existing one if set to disabled.
+func (env *Environment) InitPluginHealthCheckJob(enable bool) {
+	// Config is set to enable. No job exists, start a new job.
+	if enable && env.pluginHealthCheckJob == nil {
+		job := newPluginHealthCheckJob(env)
+		env.pluginHealthCheckJob = job
+		job.Start()
+	}
+
+	// Config is set to disable. Job exists, kill existing job.
+	if !enable && env.pluginHealthCheckJob != nil {
+		env.pluginHealthCheckJob.Cancel()
+		env.pluginHealthCheckJob = nil
+	}
+}
+
+// Start continuously runs health checks on all active plugins, on a timer.
+func (job *PluginHealthCheckJob) Start() {
+	interval := time.Duration(HEALTH_CHECK_INTERVAL) * time.Second
+	mlog.Debug(fmt.Sprintf("Plugin health check job starting. Sending health check pings every %v seconds.", interval))
+
+	go func() {
+		defer close(job.cancelled)
+
+		ticker := time.NewTicker(interval)
+		defer func() {
+			ticker.Stop()
+		}()
+
+		for {
+			select {
+			case <-ticker.C:
+				activePlugins := job.env.Active()
+				for _, plugin := range activePlugins {
+					job.checkPlugin(plugin.Manifest.Id)
+				}
+			case <-job.cancel:
+				return
+			}
+		}
+	}()
+}
+
+// checkPlugin determines the plugin's health status, then handles the error or success case.
+func (job *PluginHealthCheckJob) checkPlugin(id string) {
+	var ap activePlugin
+	if p, ok := job.env.activePlugins.Load(id); ok {
+		ap = p.(activePlugin)
+	} else {
+		return
+	}
```

Comment:
Nit: You can drop the else here:

```suggestion
	p, ok := job.env.activePlugins.Load(id)
	if !ok {
		return
	}
	ap := p.(activePlugin)
```